Mathlib/NumberTheory/ModularForms/EisensteinSeries/Basic.lean

```diff
@@ -0,0 +1,119 @@
+/-
+Copyright (c) 2024 Chris Birkbeck. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Birkbeck, David Loeffler
+-/
+import Mathlib.NumberTheory.ModularForms.SlashInvariantForms
+import Mathlib.NumberTheory.ModularForms.CongruenceSubgroups
+
+/-!
+# Eisenstein Series
+
+## Main definitions
+
+* We define Eisenstein series of level `Γ(N)` for any `N : ℕ` and weight `k : ℤ` as the infinite sum
+  `∑' v : (Fin 2 → ℤ), (1 / (v 0 * z + v 1) ^ k)`, where `z : ℍ` and `v` ranges over all pairs of
+  coprime integers congruent to a fixed pair `(a, b)` modulo `N`.
+* We show that they define a slash invariant form of level `Γ(N)` and weight `k`.
+
+## References
+* [F. Diamond and J. Shurman, *A First Course in Modular Forms*][diamondshurman2005]
+-/
+
+noncomputable section
+
+open ModularForm UpperHalfPlane Complex Matrix
+
+open scoped MatrixGroups
+
+namespace EisensteinSeries
+
+variable (N : ℕ) (a : Fin 2 → ZMod N)
+
+section gammaSet_def
+
+/-- The set of pairs of coprime integers congruent to `a` mod `N`. -/
+def gammaSet := {v : Fin 2 → ℤ | (↑·) ∘ v = a ∧ IsCoprime (v 0) (v 1)}
+
+lemma pairwise_disjoint_gammaSet : Pairwise (Disjoint on gammaSet N) := by
+  refine fun u v huv ↦ ?_
+  contrapose! huv
+  obtain ⟨f, hf⟩ := Set.not_disjoint_iff.mp huv
+  exact hf.1.1.symm.trans hf.2.1
+
+/-- For level `N = 1`, the gamma sets are all equal. -/
+lemma gammaSet_one_eq (a a' : Fin 2 → ZMod 1) : gammaSet 1 a = gammaSet 1 a' :=
+  congr_arg _ (Subsingleton.elim _ _)
+
+/-- For level `N = 1`, the gamma sets are all equivalent; this is the equivalence -/
+def gammaSet_one_equiv (a a' : Fin 2 → ZMod 1) : gammaSet 1 a ≃ gammaSet 1 a' :=
+  Equiv.Set.ofEq (gammaSet_one_eq a a')
+
+end gammaSet_def
+
+variable {N a}
+
+section gamma_action
+
+/-- Right-multiplying by `γ ∈ SL(2, ℤ)` sends `gammaSet N a` to `gammaSet N (vecMul a γ)`. -/
+lemma vecMul_SL2_mem_gammaSet {v : Fin 2 → ℤ} (hv : v ∈ gammaSet N a) (γ : SL(2, ℤ)) :
+    vecMul v γ ∈ gammaSet N (vecMul a γ) := by
+  refine ⟨?_, hv.2.vecMulSL γ⟩
+  have := RingHom.map_vecMul (m := Fin 2) (n := Fin 2) (Int.castRingHom (ZMod N)) γ v
+  simp only [eq_intCast, Int.coe_castRingHom] at this
+  simp_rw [Function.comp, this, hv.1]
+  rfl
```

Comment:
This is also solved by
```suggestion
  simp
```
which doesn't depend on the exact definitions involved